calico/felix/ipsets.py

```diff
@@ -289,13 +289,30 @@ def __init__(self, tag, ip_type):
         # Members which really are in the ipset.
         self.programmed_members = None
 
-        # Do the sets exist?
-        self.set_exists = ipset_exists(self.name)
-        self.tmpset_exists = ipset_exists(self.tmpname)
+        # Do the sets exist? We don't know yet.
+        self.set_exists = None
+        self.tmpset_exists = None
 
         # Notified ready?
         self.notified_ready = False
 
+    def _set_exists(self, real):
+        """
+        Returns whether the set exists. We only check once and remember, but
+        must not do so until the greenlet is running (in case another greenlet
+        for this ipset is still working under our feet).
+
+        :param bool real: Whether the real set (vs. the tmp one).
+        """
+        if real:
+            if self.set_exists is None:
+                self.set_exists = ipset_exists(self.name)
+            return self.set_exists
+        else:
+            if self.tmpset_exists is None:
+                self.tmpset_exists = ipset_exists(self.tmpname)
+            return self.tmpset_exists
+
```

Comment:
Think I'd prefer passing in the name of the set to check for and caching in a dict to avoid that duplication.